### ISSUE TITLE
Sync Error HashState loading into HashedStorage

### ISSUE THREAD

**Issue by sglyon** [ledgerwatch/erigon#2977]
#### System information

Erigon version: `./erigon --version`

erigon version 2021.11.3-beta

OS & Version: Linux Docker

#### Expected behaviour

Trying to sync. Should work

#### Actual behaviour

Sync always gets error on HashState step when loading HashedStorage

Logs here:  [erigon_logs.log](https://github.com/ledgerwatch/erigon/files/7555531/erigon_logs.log)

**Comment by AlexeyAkhunov** [comment]
Thanks for reporting. This is one of the errors we dread the most (DB corruption).
I would advise to reset the hashState stage to make it work. First, shut down Erigon. Then:
```
./build/bin/integration stage_hash_state --reset --datadir <your_datadir>
./build/bin/integration stage_trie --reset --datadir <your_datadir>
```

and then start Erigon.

I am just testing it on a copy of my node to see it if works, so if you want to be sure, please wait for me to report back

**Comment by AlexeyAkhunov** [comment]
I got through the `integration` invocations and now my Erigon node is recreating hash state. I think you can try this now

**Comment by sglyon** [comment]
ok thank you!

**Comment by sglyon** [comment]
That fixed it! 

Thank you!

**Comment by sglyon** [comment]
Unfortunately this has happened again :(

I just tried the `integration` steps above, but they also fail.

Below is the log for the sync failure:
[erigon_logs.log](https://github.com/ledgerwatch/erigon/files/7722843/erigon_logs.log)

**Comment by sglyon** [comment]
the error I get when tryin gto reset the hash_state

```
./build/bin/integration stage_hash_state --reset --datadir XXX

INFO[12-17|12:52:04.960] Initialised chain configuration          config="{ChainID: 1 Homestead: 1150000 DAO: 1920000 DAOSupport: true EIP150: 2463000 EIP155: 2675000 EIP158: 2675000 Byzantium: 4370000 Constantinople: 7280000 Petersburg: 7280000 Istanbul: 9069000, Muir Glacier: 9200000, Berlin: 12244000, London: 12965000, Arrow Glacier: 13773000, Engine: ethash}"
integration: mdbx:23366: mdbx_drop_tree: Assertion `(node_flags(node) & (0x01 | 0x02 | 0x04)) == 0' failed.
SIGABRT: abort
PC=0x7fd9774dbd22 m=21 sigcode=18446744073709551610
signal arrived during cgo execution

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0xe0eb30, 0xc0007959e0)
        runtime/cgocall.go:156 +0x5c fp=0xc0007959b8 sp=0xc000795980 pc=0x43937c
github.com/torquem-ch/mdbx-go/mdbx._Cfunc_mdbx_drop(0x2e2c010, 0x14, 0x0)
        _cgo_gotypes.go:585 +0x4c fp=0xc0007959e0 sp=0xc0007959b8 pc=0x8eae6c
github.com/torquem-ch/mdbx-go/mdbx.(*Txn).Drop.func1(0xc0016c04c0, 0x1bd6b00, 0x0)
        github.com/torquem-ch/mdbx-go@v0.22.1/mdbx/txn.go:472 +0x5c fp=0xc000795a20 sp=0xc0007959e0 pc=0x8f4d5c
github.com/torquem-ch/mdbx-go/mdbx.(*Txn).Drop(0x1013d20, 0x70c000, 0xc0)
        github.com/torquem-ch/mdbx-go@v0.22.1/mdbx/txn.go:472 +0x1d fp=0xc000795a50 sp=0xc000795a20 pc=0x8f4bdd
github.com/ledgerwatch/erigon-lib/kv/mdbx.(*MdbxTx).ClearBucket(0xc00704c3c0, {0x11454c0, 0xc0000f0200})
        github.com/ledgerwatch/erigon-lib@v0.0.0-20211104110507-597d0fbb01ab/kv/mdbx/kv_mdbx.go:700 +0x66 fp=0xc000795a80 sp=0xc000795a50 pc=0x8fb026
github.com/ledgerwatch/erigon/eth/stagedsync.ResetHashState({0x1bd6b00, 0xc00704c3c0})
        github.com/ledgerwatch/erigon/eth/stagedsync/stage_interhashes.go:465 +0x37 fp=0xc000795ab8 sp=0xc000795a80 pc=0xdbf617
github.com/ledgerwatch/erigon/cmd/integration/commands.stageHashState({0x1bbb060, 0xc000708000}, {0x1bb3df8, 0xc00051e200})
        github.com/ledgerwatch/erigon/cmd/integration/commands/stages.go:682 +0x219 fp=0xc000795cd0 sp=0xc000795ab8 pc=0xdfc879
github.com/ledgerwatch/erigon/cmd/integration/commands.glob..func13(0x22d1cc0, {0x113c5db, 0x3, 0x3})
        github.com/ledgerwatch/erigon/cmd/integration/commands/stages.go:136 +0xf1 fp=0xc000795da0 sp=0xc000795cd0 pc=0xe09ef1
github.com/spf13/cobra.(*Command).execute(0x22d1cc0, {0xc0005094a0, 0x3, 0x3})
        github.com/spf13/cobra@v1.2.1/command.go:856 +0x60e fp=0xc000795e60 sp=0xc000795da0 pc=0x5d8c0e
github.com/spf13/cobra.(*Command).ExecuteC(0x22d4c40)
        github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bc fp=0xc000795f18 sp=0xc000795e60 pc=0x5d92fc
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.2.1/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.2.1/command.go:895
main.main()
        github.com/ledgerwatch/erigon/cmd/integration/main.go:15 +0x112 fp=0xc000795f80 sp=0xc000795f18 pc=0xe0e1b2
runtime.main()
        runtime/proc.go:255 +0x227 fp=0xc000795fe0 sp=0xc000795f80 pc=0x46df07
runtime.goexit()
        runtime/asm_amd64.s:1581 +0x1 fp=0xc000795fe8 sp=0xc000795fe0 pc=0x49f041

goroutine 6 [sleep]:
time.Sleep(0x22ecb25c00)
        runtime/time.go:193 +0x12e
github.com/VictoriaMetrics/metrics.summariesSwapCron(0x45d964b800)
        github.com/VictoriaMetrics/metrics@v1.18.1/summary.go:237 +0x36
created by github.com/VictoriaMetrics/metrics.registerSummaryLocked
        github.com/VictoriaMetrics/metrics@v1.18.1/summary.go:211 +0x127

goroutine 8 [sync.Cond.Wait]:
sync.runtime_notifyListWait(0xc0002e6048, 0x0)
        runtime/sema.go:513 +0x13d
sync.(*Cond).Wait(0xc000351620)
        sync/cond.go:56 +0x8c
github.com/anacrolix/stm.(*Tx).wait(0xc0002e6000)
        github.com/anacrolix/stm@v0.2.1-0.20201002073511-c35a2c748c6a/tx.go:71 +0xb8
github.com/anacrolix/stm.Atomically(0xc0675171321b4900)
        github.com/anacrolix/stm@v0.2.1-0.20201002073511-c35a2c748c6a/funcs.go:80 +0x14f
github.com/anacrolix/stm/rate.(*Limiter).tokenGenerator(0xc00032a960, 0x2625a00)
        github.com/anacrolix/stm@v0.2.1-0.20201002073511-c35a2c748c6a/rate/ratelimit.go:59 +0x170
created by github.com/anacrolix/stm/rate.NewLimiter
        github.com/anacrolix/stm@v0.2.1-0.20201002073511-c35a2c748c6a/rate/ratelimit.go:45 +0x229

goroutine 9 [chan receive]:
github.com/anacrolix/dht/v2.dnsResolverRefresher()
        github.com/anacrolix/dht/v2@v2.8.0/dht.go:103 +0x70
created by github.com/anacrolix/dht/v2.init.0
        github.com/anacrolix/dht/v2@v2.8.0/dht.go:109 +0x25

goroutine 10 [select]:
github.com/ledgerwatch/erigon/cmd/utils.RootContext.func1()
        github.com/ledgerwatch/erigon/cmd/utils/cmd.go:88 +0x1ae
created by github.com/ledgerwatch/erigon/cmd/utils.RootContext
        github.com/ledgerwatch/erigon/cmd/utils/cmd.go:79 +0xa7

goroutine 11 [select]:
github.com/ledgerwatch/erigon/internal/debug.ListenSignals({0x0, 0x0})
        github.com/ledgerwatch/erigon/internal/debug/signal.go:26 +0x1e5
created by github.com/ledgerwatch/erigon/internal/debug.SetupCobra
        github.com/ledgerwatch/erigon/internal/debug/flags.go:173 +0x1ee

goroutine 12 [select]:
github.com/ledgerwatch/erigon/cmd/utils.RootContext.func1()
        github.com/ledgerwatch/erigon/cmd/utils/cmd.go:88 +0x1ae
created by github.com/ledgerwatch/erigon/cmd/utils.RootContext
        github.com/ledgerwatch/erigon/cmd/utils/cmd.go:79 +0xa7

goroutine 53 [syscall]:
os/signal.signal_recv()
        runtime/sigqueue.go:169 +0x98
os/signal.loop()
        os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
        os/signal/signal.go:151 +0x2c

rax    0x0
rbx    0x7dd8f3fff640
rcx    0x7fd9774dbd22
rdx    0x0
rdi    0x2
rsi    0x7dd8f3ffe820
rbp    0x7fd9776618f0
rsp    0x7dd8f3ffe820
r8     0x0
r9     0x7dd8f3ffe820
r10    0x8
r11    0x246
r12    0x1bfd32f
r13    0x5b46
r14    0x1c041e8
r15    0x7dd8ec000f00
rip    0x7fd9774dbd22
rflags 0x246
cs     0x33
fs     0x0
gs     0x0

```

**Comment by AskAlexSharov** [comment]
Likely you have hardware problem - disk or ram. Try https://www.memtest86.com/

**Comment by sglyon** [comment]
Are you thinking faulty hardware, or insufficient hardware? I think I should have "enough" hardware -- 8 TB SSD and 32 GB RAM. 

I haven't used memtest86 before. Would it be testing bad ram?

**Comment by AskAlexSharov** [comment]
Yes, i mean broken hardware

**Comment by AskAlexSharov** [comment]
```
make db-tools

build/bin/mdbx_stat -ef <your_datadir>/chaindata

build/bin/mdbx_chk -vvv <your_datadir>/chaindata
```

**Comment by sglyon** [comment]
Thanks for continued support. 


I ran both of these commands. 

The first one (stat) gave the following output:

```
mdbx_stat v0.11.1-26-ge93d53e (2021-11-21T02:38:14+03:00, T-c00e8455b6b4cd30522cf1a7a08d87625d7f8561)
Running for /mnt/samsung8tb/ethereum/erigon/chaindata...
Environment Info
  Pagesize: 4096
  Dynamic datafile: 12288..2199023255552 bytes (+2147483648/-4294967296), 3..536870912 pages (+524288/-1048576)
  Current mapsize: 2199023255552 bytes, 536870912 pages 
  Current datafile: 1812476198912 bytes, 442499072 pages
  Last transaction ID: 613801
  Latter reader transaction ID: 613801 (0)
  Max readers: 118
  Number of reader slots uses: 1
Garbage Collection
  Pagesize: 4096
  Tree depth: 2
  Branch pages: 1
  Leaf pages: 18
  Overflow pages: 9088
  Entries: 711
Page Usage
  Total: 536870912 100%
  Backed: 442499072 82.4%
  Allocated: 442484536 82.4%
  Remained: 94386376 17.6%
  Used: 433227365 80.7%
  GC: 9257171 1.7%
  Retained: 228209 0.0%
  Reclaimable: 9028962 1.7%
  Available: 103415338 19.3%
Status of Main DB
  Pagesize: 4096
  Tree depth: 1
  Branch pages: 0
  Leaf pages: 1
  Overflow pages: 0
  Entries: 51
```

I let the other command run for 10 minutes, the output ended up being over 400 MB before I stopped it

**Comment by AskAlexSharov** [comment]
Stat looks ok. 
reduce amount of -vvvv to reduce output size

**Comment by sglyon** [comment]
Took a little while to run, but appears to have found some issues:

```
mdbx_chk v0.11.1-26-ge93d53e (2021-11-21T02:38:14+03:00, T-c00e8455b6b4cd30522cf1a7a08d87625d7f8561)
Running for /mnt/samsung8tb/ethereum/erigon/chaindata in 'read-only' mode...
 - monopolistic mode
 - current boot-id 9ebf8d5de3353b7a-11d41b3a087344ae
 - pagesize 4096 (4096 system), max keysize 1980..2022, max readers 118
 - mapsize 2199023255552 (2.00 Tb)
 - dynamic datafile: 12288 (12.00 Kb) .. 2199023255552 (2.00 Tb), +2147483648 (2.00 Gb), -4294967296 (4.00 Gb)
 - current datafile: 1812476198912 (1.65 Tb), 442499072 pages
 - meta-0: steady txn#613801, head
 - meta-1: steady txn#613799, tail
 - meta-2: steady txn#613800, stay
 - transactions: recent 613801, latter reader 613801, lag 0
Traversal b-tree by txn#613801...
 - found 'AccountChangeSet' area
 - found 'AccountHistory' area
 - found 'BlockBody' area
 - found 'BlockTransaction' area
 - found 'BlockTransactionLookup' area
 - found 'CallFromIndex' area
 - found 'CallToIndex' area
 - found 'CallTraceSet' area
 - found 'CanonicalHeader' area
 - found 'Code' area
 - found 'Config' area
 - found 'DbInfo' area
 - found 'HashedAccount' area
 ! corrupted branch-page #325501653, mod-txnid 595722
 ! node[75] key (10491) beyond page-end
 - found 'HashedCodeHash' area
 - found 'HashedStorage' area
 - found 'Header' area
 - found 'HeaderNumber' area
 - found 'HeadersTotalDifficulty' area
 - found 'IncarnationMap' area
 - found 'LastBlock' area
 - found 'LastHeader' area
 - found 'LogAddressIndex' area
 - found 'LogTopicIndex' area
 - found 'Migration' area
 - found 'PlainCodeHash' area
 - found 'PlainState' area
 - found 'Receipt' area
 - found 'Sequence' area
 - found 'StorageChangeSet' area
 - found 'StorageHistory' area
 - found 'SyncStage' area
 - found 'TransactionLog' area
 - found 'TxSender' area
 - problems: invalid/corrupted (1)
 - pages: walked 433227281, left/unused 9257255
 - summary: average fill 77.4%, 1 problems
Skip processing @MAIN since tree is corrupted (1 problems)
Processing '@GC'...
 - key-value kind: ordinal-key => single-value
 - last modification txn#613801
 - fixed key-size 8
 - summary: 711 records, 0 dups, 5688 key's bytes, 37031528 data's bytes, 0 problems
 - space: 536870912 total pages, backed 442499072 (82.4%), allocated 442484536 (82.4%), available 103415338 (19.3%)
Total 3 errors are detected, elapsed 60060.701 seconds.
```
